File: ejercicio/datos_alumnos.py
datos = []

def cargar_alumnos():
    nombre = input("Ingrese el nombre del alumno: ")

    alumno = next((al for al in datos if al[0] == nombre), None)
    if alumno:
        print("El alumno ya existe")
        opcion = input("¿Quiere agregar (A) o modificar (M) una nota?: ").upper()

        if opcion == "A":
            nueva_materia = input("Ingrese el nombre de la nueva materia: ")
            nueva_nota = int(input("Ingrese la nota: "))
            alumno[1].append([nueva_materia, nueva_nota])
            print(f"Se agregó la materia {nueva_materia} con la nota {nueva_nota}")
        elif opcion == "M":
            materia = input("Ingrese el nombre de la materia a modificar: ")
            nota = int(input("Ingrese la nueva nota: "))
            materia_encontrada = next((m for m in alumno[1] if m[0] == materia), None)

            if materia_encontrada:
                materia_encontrada[1] = nota
                print(f"Se actualizó la materia {materia} con la nota {nota}")
            else:
                print("Materia no encontrada")
        else:
            print("Opción no válida")
    else:
        materias = []
        while True:
            materia = input("Ingrese el nombre de la materia: ")
            nota = int(input("Ingrese la nota de la materia: "))
            materias.append([materia, nota])
            continuar = input("¿Desea agregar otra materia? (s/n): ").lower()
            if continuar != 's':
                break
        datos.append([nombre, materias])
        print("Alumno agregado correctamente")

while input("¿Desea agregar datos de un alumno? (s/n): ").lower() == 's':
    cargar_alumnos()

print("\nDatos de alumnos:")
for alumno in datos:
    nombre = alumno[0]
    print(f"{nombre}:")
    for materia in alumno[1]:
        print(f" - {materia[0]}: {materia[1]}")
    print("")

File: ejercicio/datos_alumnos.js
const readlineSync = require('readline-sync');

const datos = [];

function cargarAlumnos() {
    let nombre = readlineSync.question("Ingrese el nombre del alumno: ");

    let alumno = datos.find(al => al[0] === nombre);
    if (alumno){
        console.log("El alumno ya existe");
        let opcion = readlineSync.question("¿Quiere agregar (A) o modificar (M) una nota?: ").toUpperCase();

    if (opcion === "A"){
        let nuevaMateria = readlineSync.question("Ingrese el nombre de la nuevaMateria: ");
        let nuevaNota = parseInt(readlineSync.question("Ingrese la nota: "));
        alumno[1].push([nuevaMateria, nuevaNota]);
        console.log(`Se agregó la materia ${nuevaMateria} con la nota ${nuevaNota}`);
    } 
    else if (opcion === "M"){
        let materia = readlineSync.question("Ingrese el nombre de la materia a modificar: ");
        let nota = parseInt(readlineSync.question("Ingrese la nueva nota: "));
        let materiaEncontrada = alumno[1].find(m => m[0] === materia);

        if(materiaEncontrada){
            materiaEncontrada[1] = nota;
            console.log(`Se actualizó la materia ${materia} con la nota ${nota}`);
        } else {
        console.log("Matería no encontrada");
        }
    } else {
        console.log("Opción no válida");
        }
    } else {
        let materias = [];
        let continuar = true;

        while(continuar){
            let materia = readlineSync.question("Ingrese el nombre de la materia: ");
            let nota = parseInt(readlineSync.question("Ingrese la nota de la materia: "));
            materias.push([materia, nota]);
            continuar = readlineSync.keyInYN("Desea agregar otra materia?");
        }
        datos.push([nombre, materias]);
        console.log("Alumno agregado correctamente");
    }
}

while(readlineSync.keyInYN("Desea agregar datos de un alumno?")){
    cargarAlumnos();
}

console.log("\n Datos de alumnos:");
datos.forEach(alumno => {
    let nombre = alumno[0];
    console.log(`${nombre}:`);
    alumno[1].forEach(materia => {
        console.log(` - ${materia[0]}: ${materia[1]}`);
    });
    console.log("");
});